include/openssl/opensslv.h

```diff
@@ -10,92 +10,126 @@
+/* Could be: #define OPENSSL_VERSION_BUILD_METADATA "+vendor.1" */
+# undef OPENSSL_VERSION_BUILD_METADATA
+
+/*
+ * Note: OPENSSL_VERSION_BUILD_METADATA will never be defined by
+ * the OpenSSL Project, it's entirely reserved for others vendors
+ */
+
+/*
+ * Shared library version
  *
- * So, here's the way it works here: first of all, the library version
- * number doesn't need at all to match the overall OpenSSL version.
- * However, it's nice and more understandable if it actually does.
- * The current library version is stored in the macro SHLIB_VERSION_NUMBER,
- * which is just a piece of text in the format "M.m.e" (Major, minor, edit).
- * For the sake of Tru64, IRIX, and any other OS that behaves in similar ways,
- * we need to keep a history of version numbers, which is done in the
- * macro SHLIB_VERSION_HISTORY.  The numbers are separated by colons and
- * should only keep the versions that are binary compatible with the current.
+ * This is strictly to express ABI version, which may or may not
+ * be related to the API version expressed with the macros above.
+ * This is defined in free form.
+ */
+# define OPENSSL_SHLIB_VERSION 2
+
+/*
+ * SECTION 2: USEFUL MACROS AND FUNCTIONS
+ */
+
+/* For checking general API compatibility when preprocessing */
+# define OPENSSL_VERSION_PREREQ(maj,min)                                \
+1    ((OPENSSL_VERSION_MAJOR << 16) + OPENSSL_VERSION_MINOR >= (maj << 16) + min)
+
+/*
+ * These return the values of OPENSSL_VERSION_MAJOR, OPENSSL_VERSION_MINOR,
+ * OPENSSL_VERSION_PATCH, OPENSSL_VERSION_PRE_RELEASE and
+ * OPENSSL_VERSION_BUILD_METADATA, respectively.
+ */
+unsigned int OPENSSL_version_major(void);
+unsigned int OPENSSL_version_minor(void);
+unsigned int OPENSSL_version_patch(void);
+const char *OPENSSL_version_pre_release(void);
+const char *OPENSSL_version_build_metadata(void);
+
+/*
+ * Returns the string "MAJOR.MINOR.PATCH", where MAJOR, MINOR and PATCH
+ * are replaced with the values from the corresponding OPENSSL_VERSION_
+ * macros.
+ */
+const char *OPENSSL_version_text(void);
+/*
+ * Returns the string that concatenates what OPENSSL_version_text returns
+ * together with OPENSSL_VERSION_PRE_RELEASE and OPENSSL_VERSION_BUILD_METADATA.
+ */
+const char *OPENSSL_version_text_full(void);
+
+/*
+ * SECTION 3: ADDITIONAL METADATA
  */
-# define SHLIB_VERSION_HISTORY ""
-# define SHLIB_VERSION_NUMBER "1.1"
+# define OPENSSL_RELEASE_DATE "xx XXX xxxx"
+
+/*
+ * SECTION 3: BACKWARD COMPATIBILITY
+ */
+# include <openssl/opensslconf.h>
+
+# if OPENSSL_API_COMPAT < 0x40000000L
+#  ifdef OPENSSL_NO_NEWVERSION_ONLY
+/* Helper functions */
+#   define OPENSSL_MSTR_HELPER(x) #x
+#   define OPENSSL_MSTR(x) OPENSSL_MSTR_HELPER(x)
 
+/* Synthesize OPENSSL_VERSION_NUMBER with the layout 0xMNN00PPSL */
+#   define OPENSSL_VERSION_NUMBER        \
+    (long)( (OPENSSL_VERSION_MAJOR<<28)  \
+            |(OPENSSL_VERSION_MINOR<<20) \
+            |(OPENSSL_VERSION_PATCH<<4)  \
+            |0x0 )
```

Comment:
So I understand that this would turn it into 0x30000000, which means that maximum version we can go to is 15. Maybe we should consider something else, but it should clearly be larger than the current number, or deprecating this define.

Reply:
Or we could go for a 64 bit version number ...

Reply:
Correct. OPENSSL_VERSION_NUMBER is considered deprecated usage.

Reply:
Oh, I didn't see the if OPENSSL_API_COMPAT < 0x40000000L part.